### PR TITLE
Add AbstractOrderedDict

### LINKED ISSUE #38141: AbstractOrderedDict (and AbstractOrderedSet) in Base
Hi, we probably need AbstractOrderedDict <: AbstractDict, in Julia 1.6, even if my Dict changed to ordered doesn't make it into Julia 1.6.

Right now AbstractDict doesn't require order, and maybe later we should change that (after packages refer to the new type).

I'm not sure we need bith right away:

AbstractOrderedDict <: AbstractUnorderedDict

And I'm not changes Set, for now at least, but I guess we should similarly have AbstractOrderedSet.

### PR DESCRIPTION
**Pull request by PallHaraldsson**
Fixes #38141 regarding it, and if people ok with this (discussion there), then similar AbstractOrderedSet should go in.

### CODE REVIEW

File: base/essentials.jl [hunk 1]
```diff
@@ -17,12 +17,21 @@ abstract type AbstractSet{T} end
...
+and values of type `V`. [`Dict`](@ref), [`IdDict`](@ref) and other types
+are subtypes of this. An `AbstractDict{K, V}` should be an iterator of `Pair{K, V}`.
 """
 abstract type AbstractDict{K,V} end
 
+"""
+    AbstractOrderedDict{K, V}
+
+Supertype for ordered dictionary-like types with keys of type `K` and values of
+type `V`. [`Dict`](@ref), [`IdDict`](@ref) and other types are subtypes of this.
```
Comment (JeffBezanson) [inline base/essentials.jl:30]: They're not?

Comment (PallHaraldsson) [reply base/essentials.jl:30]: Ok, sorry you're right, a bit premature... They're only ordered on my other PR (I thought it was on the finish-line, but it maight not make it). You like be to skip that part to merge, or wait for my other PR to be merged first?

**Comment by simeonschaub** [comment]
I wonder whether a trait would make more sense here, since that would make implementing Dict wrapper types easier and overall would be a bit more extensible.

**Comment by PallHaraldsson** [comment]
@simeonschaub, I shorteded my doc test, after Jeff's (now outdated) comment, so this PR COULD be merged. I'm however not sure about traits and your idea, so you or someone would have to show the way, if you could please if there's an alternative.

I at leas wanted to leave with this one mergeable, as it seems good to me independen of if I succeed with my other PR[s] for ordered Dict (and maybe for Set).

The tests on the main platforms e.g. linux, were successful, only failing on alternative platforms e.g. linuxaarch64, unrelated to my change:

```
Error in testset Distributed:
Error During Test at /buildworker/worker/tester_linuxaarch64/build/share/julia/test/testdefs.jl:21
  Got exception outside of a @test
  LoadError: Distributed test failed, cmd : `/buildworker/worker/tester_linuxaarch64/build/bin/julia -Cnative -J/buildworker/worker/tester_linuxaarch64/build/lib/julia/sys.so -g1 --startup-file=no --check-bounds=yes --startup-file=no --depwarn=error /buildworker/worker/tester_linuxaarch64/build/share/julia/stdlib/v1.6/Distributed/test/distributed_exec.jl`
  Stacktrace:
```

I'm getting a very frustrated (and frustrating others) with such false alarms, and real alarms (elsewhere). I thought this change at least would be easy, and actually the my main PR (and it was, in a way, except for tests), I had no idea how deep the rabbit hole was.

**Comment by rfourquet** [comment]
> I wonder whether a trait would make more sense here

I have been thinking about the possibility of adding a supertype to `AbstractDict`, `Associative`, such that `AbstractMultiDict` could inherit from it. Many functions accepting a dictionary would work equally well with a multidict (a dict which can have multiple entries associated with the same key). With this scheme, it might indeed be more practical to have "ordered" as a trait. Otherwise, you end-up with:
```
AbstractOrderedDict <: AbstractDict <: Associative
AbstractOrderedMultiDict <: AbstractMultiDict <: Associative
AbstractOrderedAssociative = Union{AbstractOrderedDict, AbstractOrderedMultiDict}
```
Which is too ugly.

**Comment by tpapp** [comment]
Yes, an associative mapping can have many traits, including

- being sorted
- having a well-defined ordering
- having a default element

https://github.com/JuliaCollections/DataStructures.jl has examples for all of these.

Since organizing these into a branching type hierarchy gets tedious as outlined by @rfourquet above, these should be traits.

Also, I am not sure why these should live in `Base`. At least for the experimental phase, I would suggest a `DictBase.jl` package that just exports these traits and defines them for `Base.Dict`.

@PallHaraldsson, please also consider that nontrivial changes to the type hiearchy in `Base` usually involve more than 2 days of discussion after opening an issue, because SemVer requires that they continue to be supported in that form at least until the next major release, and in practice [much longer](https://discourse.julialang.org/t/psa-julia-is-not-at-that-stage-of-development-anymore/44872).

**Comment by PallHaraldsson** [comment]
@tpapp I was following your advise: "Generally, if you want something changed in Base and you want concrete feedback, it is best to make a PR".

I changed to Draft, I'm a bit new to using it, but it doesn't mean I'm opposed to seeing this merged, nor am I pushing for it (this version). With "this PR COULD be merged", I meant, I think so, not "you SHOULD".

> I would suggest a `DictBase.jl` package

That seems bad. We can use AbstractDict already and it's in Base and having that in only is already a problem (whether or not my new ODict proposal goes into 1.6). People already are reluctant to add OrderedCollections for only OrderedDict, and requiring people to also add DictBase and not seeming AbstractOrderedDict is Julia's docs seems bad.

@rfourquet, I admit I did not have AbstractMultiDict (or a concreate such type) in mind, but what's the actual problem with your "ugly" solution? And I'm curious how the alternative would look.

**Comment by rfourquet** [comment]
> but what's the actual problem with your "ugly" solution?

The problem is that the current inheritance mechanism doesn't support inheriting multiple supertypes at the same time, like "being ordered" and being "multidict" for an associative for example. You can kinda poorly simulate that like with my "ugly" solution, but it means 1) that you don't have one abstract "IsOrdered" type, but two or more (like `AbstractOrderedDict` and `AbstractOrderedMultiDict`  here), and that then to dispatch on "IsOrdered", you actually have to dispatch on the `AbstractOrderedAssociative` `Union`, but then it's a closed (non-extensible) system, i.e. if DataStructures.jl introduces `AbstractNotDictNorMultiDict <: Associative`, then `AbstractOrderedNotDictNorMultiDict` won't be part of the `AbstractOrderedAssociative` `Union`.

So if it's not crucial to have `Ordered` as an abstract supertype of our containers, using traits can work better as a solution, because only one trait is sufficient, e.g. 
```
foo(x::Associative) = isordered(x) ? dothis(x) : dothat(x)
```

**Comment by PallHaraldsson** [comment]
FYI: AbstractOrderedDict is not yet exported, so at least that's missing from this PR.

I also noticed, merge on two dicts (that <: AbstractDict) give you a Dict. That's overridden for OrderedDict, not not SwissDict (probably a bug there, as it seems to be an ordered one). It would be nice that the logic could be done once (here in Base?). Also merge(a, b) gets you the "a" type, OrderedDict, but not b is the OrderedDict. I'm not sure that's intentional.

**Comment by StefanKarpinski** [comment]
Why does this need to be defined in Base?

**Comment by PallHaraldsson** [comment]
When I was looking into adding/changing Base to ordered Dict, this seemed (more) sensible to have.

It still seems sensible to have this (or something similar), e.g. AbstractOrderedDict, and yes it could be in a package. It's just unclear then if in OrderedCollections or elsewhere, with ordered Dict types defined there and in two other packages.

**Comment by ViralBShah** [comment]
Are we going to accept this in Base? If not, should we close?

**Comment by tpapp** [comment]
I don't think it should go into `Base` in this form.

A package may be the best place to experiment with traits (or other approaches) related to querying properties of `Dict` & friends — ideally coordinating with packages that need this, or implement such types. Only when that matures should we consider having it in `Base`, but it is not clear why that would be necessary, as the traits can just define the relevant description of whatever is in `Base` independently.

So I would recommend closing.

**Comment by ViralBShah** [comment]
That was my thought as well. @PallHaraldsson Hopefully this can be tried out in a package and matured.

**Comment by PallHaraldsson** [comment]
If there's a better way I'm ok with that; if traits I'm ok with that, but I'm just not up to speed on that, so I'm not sure I'll be doing that. I'm ok with the issue closed, see my argument for (something) in Base, in my last comment. It's not urgent, while seems preferrable for next Julia LTS.

**Pull request closed without merging**